pubnub-core/src/pubnub.rs

```diff
@@ -174,123 +183,104 @@ impl PubNub {
-                    tx: my_tx,
-                    rx: my_rx,
-                };
-                let their_pipe = Pipe {
-                    tx: their_tx,
-                    rx: their_rx,
-                };
-
-                (my_pipe, their_pipe)
-            };
-            *guard = Some(my_pipe);
-
-            let mut channels: ChannelMap = ChannelMap::new();
-            let listeners = channels
-                .entry(channel.to_string())
-                .or_insert_with(Default::default);
-            listeners.push(channel_tx);
-
-            // Create subscribe loop
-            debug!("Creating SubscribeLoop");
-            let subscribe_loop = SubscribeLoop::new(
-                their_pipe,
-                self.client.clone(),
-                self.origin.clone(),
-                self.agent.clone(),
-                self.subscribe_key.clone(),
-                channels,
-                ChannelMap::new(),
-            );
-
-            // Spawn the subscribe loop onto the Tokio runtime
-            tokio::spawn(subscribe_loop.run());
-
-            debug!("Waiting for long-poll...");
-            guard
-                .as_mut()
-                .unwrap()
-                .rx
-                .next()
-                .await
-                .expect("Unable to receive ready message");
-        }
-
-        Subscription {
-            name: ListenerType::Channel(channel.to_string()),
-            id,
-            tx: guard.as_ref().unwrap().tx.clone(),
-            channel: channel_rx,
-        }
+    pub async fn subscribe(&mut self, channel: &str) -> Subscription<TRuntime> {
+        let supervisor_arc_clone = self.subscribe_loop_supervisor.clone();
+        let mut supervisor_guard = supervisor_arc_clone.lock().await;
+        supervisor_guard.subscribe(self, channel).await
     }
 
     /// Set the subscribe filters.
     ///
     /// # Example
     ///
     /// ```
-    /// use pubnub::PubNub;
+    /// use pubnub_hyper::PubNub;
     ///
     /// let mut pubnub = PubNub::new("demo", "demo");
     /// pubnub.filters("uuid != JoeBob");
     /// ```
     pub fn filters(&mut self, filters: &str) {
         self.filters = Some(utf8_percent_encode(filters, NON_ALPHANUMERIC).to_string());
     }
+
+    /// Get a reference to a transport being used.
+    pub fn transport(&self) -> &TTransport {
+        &self.transport
+    }
+
+    /// Get a reference to a runtime being used.
+    pub fn runtime(&self) -> &TRuntime {
+        &self.runtime
+    }
 }
 
-impl PubNubBuilder {
-    /// Create a new `PubNubBuilder` that can configure a `PubNub` client.
+impl<TTransport, TRuntime> PubNubBuilder<TTransport, TRuntime>
+where
+    TTransport: Transport,
+    TRuntime: Runtime,
+{
+    /// Build the PubNub client to begin streaming messages.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use pubnub_hyper::PubNubBuilder;
+    ///
+    /// let pubnub = PubNubBuilder::new("demo", "demo")
+    ///     .build();
```

Comment:
Right, simply properly documenting these would eliminate the odd dependency on the `pubnub-hyper`. However, the price for that, in the current situation, would be making the example `ignore`, and write some non-buildable code that users won't be able to copy-paste and use.
On the other hand, this is the most accurate example code we can provide to the users at the moment. This is why I decided to go with this for now.
A lot of the values in this builder are not used anyhow, so we might be better rethinking the core approach and clearing things up.

Reply:
It looks like the `pubnub-core` crate is not meant to be used directly by users, anyway. So documentation would target e.g. implementers of higher level crates like what `pubnub-hyper` is to Tokio.

Reply:
This is temporary until we move the unrelated variables to transports. After that, we won't have to keep the unrelated values in the builder or in the "core" `PubNub` component altogether. We might even add a more advanced builder to the "hyper" to match the current interface.

Reply:
I propose we leave it as is for now. The builder is reexported, so it might still be useful for the users until we polish the builder. What do you think, @parasyte?

Reply:
I'd like to go over the code again before I give a definitive answer, but I'm leaning toward "yes" for now. If nothing else, maybe the `use` lines could be updated?

Reply:
I'm not sure I understand what you mean. They're not invalid - doctests are passing...
The alternative we have there is to simply remove the examples. We can also give another example and give mark the code section `ignore`, but I'd prefer if all the examples are tested - that way they won't get stale over time.

Reply:
I know the tests are valid. It just looks weird to be importing `pubnub_hyper` in the docs for `pubnub_core`. If changing the crate name in the `use` line is unacceptable, then move the doctests to the `pubnub_hyper` crate?

Reply:
We could do something like this:

```rust
    /// # use pubnub_hyper::PubNubBuilder;
    /// let pubnub = PubNubBuilder::new("demo", "demo")
    ///     .build();
```

Then it won't involve the `pubnub_hyper` crate. There's a downside though - users would most likely be confused trying to reproduce that example.

I can try some other ways to eliminate the `pubnub_hyper` crate from the doctests. My thinking was, even with the explicit `pubnub_hyper` in, it's the best example we can give to the users in the current context (no magic, copy-pastable stuff), yet it's true it's far from ideal if we have to reference another crate. So, yeah, I'll just try a few other things. :smile: